tun/tun_solaris.go

```diff
@@ -0,0 +1,398 @@
+	// the ioctls that follow. Reusing the existing one does not work.
+	tunFD, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return fmt.Errorf("could not open TUN (%s)", tunNode)
+	}
+	defer unix.Close(tunFD)
+
+	// Push the IP module onto the TUN device.
+	if err = unix.IoctlSetString(tunFD, unix.I_PUSH, "ip"); err != nil {
+		return fmt.Errorf("unable to push IP module: %v", err)
+	}
+
+	// select which ppa we're using
+	req := int(unix.IF_UNITSEL)
+	err = unix.IoctlSetPointerInt(tunFD, uint(req), ppa)
+	if err != nil {
+		return fmt.Errorf("unable to IF_UNITSEL: %v", err)
+	}
+
+	// link the tun stream to the IP multiplexor
+	// If we used I_PLINK the connection would persist after we close the ipfd
+	// which causes cleanup issues. By using I_LINK and holding ipfd open,
+	// everything gets cleaned up nicely when we close the file descriptors
+	// or when the process terminates.
+	muxid, err := unix.IoctlSetIntRetInt(tun.ipfd, unix.I_LINK, tunFD)
+	if err != nil {
+		return fmt.Errorf("unable to I_LINK: %v", err)
+	}
+
+	var l unix.Lifreq
+	if err := l.SetName(tun.name); err != nil {
+		return fmt.Errorf("failed to set name on Lifreq :%v", err)
+	}
+
+	// set the IP muxid
+	reqnum := int(unix.SIOCSLIFMUXID)
+	l.SetLifruInt(muxid)
+	if err = unix.IoctlLifreq(tun.ipfd, uint(reqnum), &l); err != nil {
+		return fmt.Errorf("unable to SIOCSLIFMUXID: %v", err)
+	}
+
+	return nil
+}
+
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+func (tun *NativeTun) Events() <-chan Event {
+	return tun.events
+}
+
+func (tun *NativeTun) Read(buffs [][]byte, sizes []int, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(tun.fd, nil, buffs[0][offset:])
+		if err != nil {
+			return 0, err
+		}
+		sizes[0] = len(read)
+		return 1, nil
+	}
+}
+
+func (tun *NativeTun) Write(buffs [][]byte, offset int) (int, error) {
+	for i, buf := range buffs {
+		buf = buf[offset:]
+		if err := unix.Putmsg(tun.fd, nil, buf, 0); err != nil {
+			return i, err
+		}
+	}
+	return len(buffs), nil
+}
+
+func (tun *NativeTun) Close() error {
+	var err1, err2, err3 error
+	tun.closeOnce.Do(func() {
+		err1 = tun.tunFile.Close()
+		err2 = unix.Close(tun.ipfd)
+		if tun.routeSocket != -1 {
+			unix.Shutdown(tun.routeSocket, unix.SHUT_RDWR)
+			err3 = unix.Close(tun.routeSocket)
+			tun.routeSocket = -1
+		} else if tun.events != nil {
+			close(tun.events)
+		}
+	})
+	// Clean up anything remaining
+	// e.g. ipv6 addresses that were configured onto the device
+	tunDestroy(tun.name)
+	if err1 != nil {
+		return err1
+	}
+	if err2 != nil {
+		return err2
+	}
+	return err3
+}
+
+func (tun *NativeTun) setMTU(n int) error {
+	var l unix.Lifreq
+	if err := l.SetName(tun.name); err != nil {
+		return err
+	}
+
+	reqnum := int(unix.SIOCSLIFMTU)
```

Comment:
reqnum doesn't need to be a separate variable.

Reply:
See also: https://go.googlesource.com/sys/+/master/unix/syscall_solaris_test.go#370

Reply:
> See also: https://go.googlesource.com/sys/+/master/unix/syscall_solaris_test.go#370

Dunno if that counts; I wrote it. lolsob

Reply:
The compiler isn't confused; you are. Your comment in that test isn't accurate and does a disservice.

It looks like ioctl has the wrong signature in Go. According to https://illumos.org/man/2/ioctl req is an int. In Go, that's int32, not int or uint as the Go code has now. Presumably it's too late to fix that, which is unfortunate.

What you can do, though, is just avoid the negative ioctl constant for the ones you added, which probably have no users yet anyway. This way you don't run into ambiguity with sign extension. The easiest way to do this is probably just by adding `& 0xffffffff` onto the end of the constant definition. While you're at it, you can fix your bogus comment in the test too.

Be sure to submit this to x/sys with a compelling argument that you understand well; don't just link to this comment or something and call it a day.